Add autoconfiguration for Spring MVC

dsyer opened the pull request:
Provisional structure of autoconfiguration and use in one of the sample. Maybe need to add more conditionals later. Also need to add WebFlux, if this is the right approach.

I had to hack the module-info.java a lot, and I'm a JPMS moron, so please help me understand any changes needed there.

dsyer commented:
I don't think I like the `SpringTemplateConfig`. I kind of get the motivation, but I don't want to have to remember to add `@Component` to all my renderers and apparently JStachio doesn't work without that. Did I miss something? I guess we could make the `SpringJStachio` conditional on actually finding some of those beans.

dsyer commented:
Oops. Your new test now fails with this autoconfiguration. I reverted to before the test was added.

agentgt commented on opt/jstachio-spring-example/src/main/java/module-info.java:
```diff
@@ -27,30 +27,23 @@
 	requires static spring.jcl;
-	requires static jakarta.servlet;
-	
-	requires spring.web;
```
Because spring cannot be `requires transitive` in any of our modules you will need to leave those `requires spring.*` in.

agentgt reviewed (changes requested):
Most of the changes just require you go and change all `requires transitive spring.*` to `requires spring.*`

Then every downstream module (like the example projects) will still need `requires spring.*`. The rule of thumb is to never `requires transitive` on any automatic module (no module-info.class).

Some day Spring will have true modules and your way would have worked. So you had the right intention.

agentgt commented:
> I fixed it like you said. But I don't understand because it did actually work.

The transitive will appear to work in most configurations but in a true modular one it will not.

It is kind of complicated why that is.

I'll find some more docs as to why that is the case later.

agentgt commented:
@dsyer see this:

https://stackoverflow.com/questions/46750408/why-does-javac-complain-about-named-automatic-modules

I will try to find more.

As a bonus albeit tedious doing it the non `transitive` way removes a compiler warning.

dsyer commented:
When I do that the new integration test fails. I don't understand why. Something to do with the `HelloModelAndView`? I can push it to my fork if you want.

agentgt commented:
Don’t worry I’ll take a look shortly and either comment a fix or do a PR to your branch

agentgt commented:
I can't figure out why but the auto configure is not picking it up.

I even converted the project to an Automatic module and it still never instantiates the `io.jstach.opt.spring.boot.webmvc.JStachioWebMvcAutoConfiguration`.

Tried both spring:boot-run and junit.

I even tried the old spring.factories way.

dsyer commented:
OK. I’ll figure it out from there.

agentgt commented:
I think I found the problem. It is because the example application is a modularized. 

I need to figure out a way to replace Spring Boot's Classloader used for autoconfiguration loading.

See org.springframework.boot.context.annotation.ImportCandidates

agentgt commented:
You had the wrong name for the imports file.

You had:

org.springframework.boot.AutoConfiguration.imports

It should be:

org.springframework.boot.autoconfigure.AutoConfiguration

I'm ashamed how long it took me to figure that out. I missed it several times and thought it was correct and thought it was a classloader issue.

dsyer commented:
Grr. Sorry. Thanks for spotting that. I would have got there eventually. Rebased and squashed with correct file name.

I still don't like the "every template is a `@Component`" pattern. Could we make that conditional on at least *some* components being templates? Or fallback to the default template finder if something isn't found that way?

dsyer commented:
Build failures seem unrelated to this change? More JPMS nonsense?

agentgt commented:
Yes I was working on that with your branch.

agentgt commented:
@dsyer pull my PR https://github.com/dsyer/jstachio/pull/2 on to your mvc-auto branch

agentgt commented:
@dsyer Maybe we should make a temporary mvc-auto branch directly on the jstachio/jstachio so that we can collaborate (no rebasing till we are done)?

For now I have been pulling your mvc-auto and rebasing on it so in theory if you pull now it should be a simple fast-forward.

dsyer commented:
With the templates it looks like you are loading from the service loader and registering each one as a singleton? Why? What if they were actually already `@Component` (then they would be registered twice)? I can't see any code that needs them to be beans. Can't they just be used as a fallback?

Unrelated: `JStachioAutoConfiguration` is now empty. We might as well put all the code from `SpringTemplateConfig` in there and delete the latter? Or am I missing something?

agentgt commented:
@dsyer When you get a chance see the update to my PR: https://github.com/dsyer/jstachio/pull/2

dsyer commented:
Merged into my fork.

agentgt commented:
@dsyer I went ahead and merge into main. Even though it probably is not completely done I think it is safe enough and I think some changes are arguably bug fixes (defaultTemplateFinder supportsType comes mind).

Feel free to pull main and submit another PR against it if you like.

I'm on vacation next week so my turn around time will be longer.

Thanks for working on this with me!
